feat(#6615): cold storage documentation (#2230)
medic/cht-docs: content/en/technical-overview/data/performance/archiving.md

---
title: "Archiving"
linkTitle: "Archiving"
weight:
description: >
  Move outdated documents from the primary database to cold storage
keywords: archiving cold-storage
relatedContent: >
  technical-overview/data/performance/purging
  technical-overview/data/db-schema

>   building/reference/api

Reviewer: For 1/2 a second I thought "oh yeah, we have got to add api docs for the new endpoint"...  😅

---

*Only available in 5.3.0 and above*

Archiving moves outdated documents from the primary `medic` database to a separate `medic-archive` database. This keeps the primary database small while the archived documents remain available for analytics and audit purposes.

Archiving differs from [purging](/technical-overview/data/performance/purging) in two important ways:

- Purging removes documents from user devices while keeping them in the `medic` database, whereas archiving removes documents from the `medic` database entirely. Archived documents are copied to `medic-archive` and then purged from `medic`, leaving no trace in the changes feed.
- Purging is conditional on user roles: the purge function decides per roles group, so a document can be purged for one user and not for another. Archiving is unconditional: an archived document is removed for all users.

> [!WARNING]

> > Archived documents are no longer available to the CHT application. They cannot be viewed in the app, used in [tasks](/building/tasks/tasks-js), [targets](/building/targets/targets-js), or contact summaries, and they are deleted from user devices on the next sync. Make sure the documents you archive are no longer needed before submitting them.

Reviewer: ```suggestion
> Archived documents are no longer available to the CHT application. They cannot be viewed in the app, used in [tasks](/building/tasks/tasks-js), [targets](/building/targets/targets-js), or contact summaries, and they are deleted from user devices on the next sync. Make sure the documents are no longer needed before archiving them.
```

- **Faster indexing.** Every document in `medic` is processed by every view and search index. With fewer documents, indexes build and update faster, both in day-to-day operation and after upgrades, when views are rebuilt from scratch.
- **Faster replication.** Calculating which documents each user can replicate is proportional to the size of the `medic` database. Removing outdated documents speeds up initial replication and routine syncs for all users.
- **Reduced disk usage.** The `medic-archive` database has no indexes, so archived documents cost only their raw storage. The same documents occupy dramatically less disk space in `medic-archive` than they did in `medic`, where every document also carries its share of view and search index storage.
- **Data is retained.** Unlike deleting documents, archiving keeps the full document, including attachments, along with an audit entry recording when it was archived. The data stays available for compliance and analytics purposes.

## How it works

Archiving is a pipeline with three stages:

> 1. You submit a list of document IDs to the [archive API endpoint](#queueing-documents-for-archiving). API validates the payload and stores archive jobs in the `medic-sentinel` database.

Reviewer: Minor, but it feels okay to be a bit generic about the queueing/job process.

```suggestion
1. You submit a list of document IDs to the [archive API endpoint](#queueing-documents-for-archiving). API validates the payload and queues the job for Sentinel.
```

> 2. Sentinel processes the queued jobs, either on a [configurable schedule](#configuration) or immediately when no schedule is configured. Each document is copied to `medic-archive` and then purged from `medic`.

Reviewer: > immediately when no schedule is configured

It is really immediate or just "within a 5m window"?  I was thinking the archiving waited until the next tick of the Sentinel schedule.

Author: Correct, it's within a 5 minute window. Thanks for the correction.

Author: Applied a change, mentioning that 'immediately' means within 5 minutes.

3. When users sync, archived documents are treated as deleted and are removed from their devices.

## Queueing documents for archiving

> Determining _which_ documents to archive is up to you. For example, you can use analytics queries to find reports older than a retention threshold, or contacts that have been inactive for years.

Reviewer: ```suggestion
Determining _which_ documents to archive is currently a manual process. For example, you can use analytics queries to find reports older than a retention threshold, or contacts that have been inactive for years.

A good place to start when considering which documents to archive is to evaluate your current purge configuration and see which documents are already marked as purged for all user roles.
```

Submit the document IDs to the archive endpoint:

```
POST /api/v1/archive
```

The request must:

- Be authenticated as a database admin.
- Use the `Content-Type: text/csv` header.
- Contain one document ID per line. Surrounding double quotes are stripped and blank lines are ignored, so a single-column CSV export works as-is.
- Be at most 32 MB. Split larger ID lists across multiple requests.

For example:

```bash
curl -X POST \
  -H "Content-Type: text/csv" \
  --data-binary @doc-ids.csv \
  https://<admin>:<password>@<host>/api/v1/archive
```

> The endpoint splits the IDs into archive job documents of at most {{< format-number 100_000 >}} IDs each, stored in the `medic-sentinel` database, and responds with `202 Accepted` and the list of created jobs:

Reviewer: ```suggestion
The endpoint splits the IDs into archive jobs and responds with `202 Accepted` and the list of created jobs:
```

```json
{
  "jobs": [
    { "id": "archive:0197c2f0-9a41-7000-8000-000000000000", "count": 100000 },
    { "id": "archive:0197c2f0-9a42-7000-8000-000000000001", "count": 25000 }
  ]
}
```

Jobs are persisted while the payload streams in, so a mid-payload failure can leave some jobs queued even though the request returns an error. Retrying the full payload is safe because archiving is idempotent; duplicate jobs only cost redundant processing.

> ## Server-side processing

Reviewer: Honestly, I think we should drop this whole section. It just feels like too much information.  This kind of info would be useful for a cht-core dev looking to extend/enhance the archive code, but I am not really sure how it would be useful to a consumer...

Author: I moved it at the bottom of the page. I think it might become relevant when people try to debug their archiving jobs.

4. Delete the document from the `medic-archive` database. This step is required: replication treats any document present in `medic-archive` as deleted, so users cannot download the restored document while it remains in the archive.

On their next sync, users with access to the document download it again with its original content.

## Considerations

- **Archiving is meant to be permanent.** Archived documents live on in `medic-archive` and [restoring one](#restoring-archived-documents) is a manual, per-document process. Do not archive documents you expect to need again.
- **Rules may break.** As with purging, any tasks or targets that depend on an archived document lose access to it. For example, archiving a report that completed a task reopens that task unless the document that created the task is also archived.
- **Users may be confused.** Documents disappear from devices without notification. Work with your users to ensure documents are only archived once there is no use for them.

> - **The `medic-archive` database grows.** Archived documents, including attachments, accumulate in `medic-archive`. Because the database has no indexes, its size is dramatically smaller than the footprint the same documents had in `medic`, but account for it in your [hosting](/hosting/) disk capacity planning all the same.

Reviewer: I think this one is unnecessary. Essentially, there is no action to take here.  Archiving docs is never going to cause _more disk space_ to be used than before. Sure, adding archiving to the mix does impact the growth curve of the size of the data on the disk, but if anything the end result should be needing to provision less disk space over time.  (And not more disk space for `medic-archive`.) 

```suggestion
```

Author: removed.
